src/firebase_functions/identity_fn.py

```diff
@@ -0,0 +1,433 @@
+    Example: '114.14.200.1'
+    """
+
+    user_agent: str
+    """
+    The user agent triggering the blocking function.
+    Example: 'Mozilla/5.0 (X11; Linux x86_64)'
+    """
+
+    additional_user_info: AdditionalUserInfo
+    """An object containing information about the user."""
+
+    credential: Credential | None
+    """An object containing information about the user's credential."""
+
+    timestamp: _dt.datetime
+    """
+    The time the event was triggered."""
+
+
+class BeforeCreateResponse(_typing.TypedDict, total=False):
+    """
+    The handler response type for 'before_user_created' blocking events.
+    """
+
+    display_name: str | None
+    """The user's display name."""
+
+    disabled: bool | None
+    """Whether or not the user is disabled."""
+
+    email_verified: bool | None
+    """Whether or not the user's primary email is verified."""
+
+    photo_url: str | None
+    """The user's photo URL."""
+
+    custom_claims: dict[str, _typing.Any] | None
+    """The user's custom claims object if available."""
+
+
+class BeforeSignInResponse(BeforeCreateResponse, total=False):
+    """
+    The handler response type for 'before_user_signed_in' blocking events.
+    """
+
+    session_claims: dict[str, _typing.Any] | None
+    """The user's session claims object if available."""
+
+
+BeforeUserCreatedCallable = _typing.Callable[[AuthBlockingEvent],
+                                             BeforeCreateResponse | None]
+"""
+The type of the callable for 'before_user_created' blocking events.
+"""
+
+BeforeUserSignedInCallable = _typing.Callable[[AuthBlockingEvent],
+                                              BeforeSignInResponse | None]
+"""
+The type of the callable for 'before_user_signed_in' blocking events.
+"""
+
+
+@_util.copy_func_kwargs(_options.BlockingOptions)
+def before_user_signed_in(
+    **kwargs,
+) -> _typing.Callable[[BeforeUserSignedInCallable], BeforeUserSignedInCallable]:
+    """
+    Handles an event that is triggered before a user is signed in.
+
+    Example:
+
+    .. code-block:: python
+
+      from firebase_functions import identity_fn
+
+      @identity_fn.before_user_signed_in()
+      def example(event: identity_fn.AuthBlockingEvent) -> identity_fn.BeforeSignInResponse | None:
+          pass
+
+    :param \\*\\*kwargs: Options.
+    :type \\*\\*kwargs: as :exc:`firebase_functions.options.BlockingOptions`
+    :rtype: :exc:`typing.Callable`
+            \\[ \\[ :exc:`firebase_functions.identity_fn.AuthBlockingEvent` \\], 
+            :exc:`firebase_functions.identity_fn.BeforeSignInResponse` \\| `None` \\]
+            A function that takes a AuthBlockingEvent and optionally returns BeforeSignInResponse.
+    """
+    options = _options.BlockingOptions(**kwargs)
+
+    def before_user_signed_in_decorator(func: BeforeUserSignedInCallable):
+        from firebase_functions.private._identity_fn import event_type_before_sign_in
```

Comment:
q: is lazy importing intentional? is there a rule of thumb you are using to choose lazy import vs top level import?

Reply:
answer: to avoid circular imports